### PR TITLE
Feature/custom validator

### PR DESCRIPTION
**Pull request by zVolodymyr**
**What kind of change does this PR introduce?**

- [ ] bugfix
- [x] feature
- [ ] refactoring / style
- [ ] build / chore
- [ ] documentation

**Did you add tests for your changes?**

- [x] Yes, my code is well tested
- [ ] Not relevant

**If relevant, did you update the documentation?**

- [x] Yes, I've updated the documentation
- [ ] Not relevant

**Summary**

I my current project I need to reject files with long names. Since validation rules may change project to project I think it's good idea to have ability to specify custom validation.

**Does this PR introduce a breaking change?**
No breaking changes

**Other information**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -615,11 +624,17 @@ export function useDropzone(options = {}) {
           files.forEach(file => {
             const [accepted, acceptError] = fileAccepted(file, accept)
             const [sizeMatch, sizeError] = fileMatchSize(file, minSize, maxSize)
-            if (accepted && sizeMatch) {
+            const customErrors = validator ? validator(file) : null;
```
Comment (rolandjitsu) [inline src/index.js:627]: Might make sense to check if `validator` is a function?

File: src/index.js [hunk 2]
```diff
@@ -615,11 +624,17 @@ export function useDropzone(options = {}) {
...
+
+            if (accepted && sizeMatch && !customErrors) {
               acceptedFiles.push(file)
             } else {
-              const errors = [acceptError, sizeError].filter(e => e)
-              fileRejections.push({ file, errors })
+              let errors = [acceptError, sizeError];
+              
+              if(customErrors)
+                errors = errors.concat(customErrors);
```
Comment (rolandjitsu) [inline src/index.js:635]:
```suggestion
              if (customErrors) {
                errors = errors.concat(customErrors);
              }
```

File: typings/react-dropzone.d.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ export interface DropzoneProps extends DropzoneOptions {
 
 export interface FileError {
   message: string;
-  code: "file-too-large" | "file-too-small"|"too-many-files"|"file-invalid-type";
+  code: "file-too-large" | "file-too-small"|"too-many-files"|"file-invalid-type"|string;
```
Comment (rolandjitsu) [inline typings/react-dropzone.d.ts:13]:
```suggestion
  code: "file-too-large" | "file-too-small" | "too-many-files" | "file-invalid-type" | string;
```

**Review by rolandjitsu — CHANGES_REQUESTED** [review]
Could you add an example of this in the examples folder (and add it to the styleguidist config as well)?

File: src/index.js (hunk 2 above)
Comment (zVolodymyr) [reply src/index.js]: updated

File: typings/react-dropzone.d.ts (hunk 3 above)
Comment (zVolodymyr) [reply typings/react-dropzone.d.ts]: updated

File: src/index.js (hunk 1 above)
Comment (zVolodymyr) [reply src/index.js:627]: I don't think it makes sense. There is TypeProps specified for this property and it should warn about incorrect values.  Also ignoring non-function values is more confusing, than getting "TypeError: ___ is not a function'. Adding such check makes sense only to throw error with better message.

Comment (rolandjitsu) [reply src/index.js:627]: I'm not saying to ignore it, I'm saying to maybe throw a more friendly message, but I guess it's ok to leave as is. As for TypeProps, do note that if ppl use this as a hook, they would get no warnings.

**Comment by zVolodymyr** [comment]
Added exmaple

**Review by rolandjitsu — APPROVED** [review]

**Comment by BartusZak** [comment]
Is there a possibility to use existing validations and add new ones as well?

I just created a custom validator and `maxFiles` and `maxSize` native validations stopped working.

@VolodymyrBaydalka

**Pull request merged**